sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java

```diff
@@ -0,0 +1,406 @@
+                    parameterDef.get(0).asVariable().asCondition(" == ", parameterDef.get(1).asVariable())
+                        .asConditionIf(ExpressionDef.constant(true).returning()),
+                    new StatementDef.DefineAndAssign(
+                        isCorrectInstance,
+                        parameterDef.get(1).asVariable().asCondition(" instanceof ", classname)
+                    ),
+                    new StatementDef.If(
+                        new ExpressionDef.Condition(" == ", isCorrectInstance, ExpressionDef.constant(false)),
+                        ExpressionDef.constant(false).returning()
+                    ),
+                    new StatementDef.DefineAndAssign(secondObject, new ExpressionDef.Cast(classname.type(), parameterDef.get(1).asVariable())),
+                    new StatementDef.DefineAndAssign(bothNullCondition, ExpressionDef.constant(false)),
+                    new StatementDef.DefineAndAssign(equalsCondition, ExpressionDef.constant(false)),
+                    new StatementDef.DefineAndAssign(isPropertyEqual, ExpressionDef.constant(false))
+                ));
+
+                // property equal checks
+                TypeDef propertyTypeDef;
+                ExpressionDef.CallInstanceMethod firstProperty;
+                ExpressionDef.CallInstanceMethod secondProperty;
+                for (PropertyElement beanProperty : properties) {
+                    propertyTypeDef = TypeDef.of(beanProperty.getType());
+                    if (beanProperty.getReadMethod().isPresent()) {
+                        firstProperty = firstObject
+                            .invoke(
+                                beanProperty.getReadMethod().get().getSimpleName(),
+                                propertyTypeDef,
+                                List.of()
+                            );
+                        secondProperty = secondObject
+                            .invoke(
+                                beanProperty.getReadMethod().get().getSimpleName(),
+                                propertyTypeDef,
+                                List.of()
+                            );
+                    } else {
+                        continue;
+                    }
+
+                    // equal check according to the properties' type
+                    if (propertyTypeDef instanceof TypeDef.Primitive) {
+                        // ==, primitives do not need null check
+                        statements.add(new StatementDef.If(
+                            new ExpressionDef.Condition(" != ", firstProperty, secondProperty),
+                            ExpressionDef.constant(false).returning()
+                        ));
+                    } else {
+                        // .equals, check for double null or equal objects
+                        statements.add(new StatementDef.Assign(bothNullCondition,
+                                new ExpressionDef.Condition(" && ",
```

Comment:
can we upstream improve the `ExpressionDef.Condition` type to make it easier to define `&&` expressions? //cc @dstepanov

Reply:
All this should be written using the fluent API, for some basic conditions we can maybe add `and`/`or`

Reply:
Should I try to do a basic implementation now or is this a thought for future? 

In my opinion, integrating parentheses into expressions and/or creating a `not` representation (`!(expression)`) would be more useful.